File: software/examples/c/example_scan_for_tags.c
#include <stdio.h>

#include "ip_connection.h"
#include "bricklet_nfc_rfid.h"

#define HOST "localhost"
#define PORT 4223
#define UID "XYZ" // Change XYZ to the UID of your NFC/RFID Bricklet

uint8_t current_tag_type = NFC_RFID_TAG_TYPE_MIFARE_CLASSIC;

// Callback function for state changed callback
void cb_state_changed(uint8_t state, bool idle, void *user_data) {
	NFCRFID *nr = (NFCRFID *)user_data;

	if(state == NFC_RFID_STATE_REQUEST_TAG_ID_READY) {
		uint8_t tag_type;
		uint8_t tid_length;
		uint8_t tid[7];

		nfc_rfid_get_tag_id(nr, &tag_type, &tid_length, tid);

		if(tid_length == 4) {
			printf("Found tag of type %d with ID [%x %x %x %x]\n", 
			       tag_type, tid[0], tid[1], tid[2], tid[3]);
		} else {
			printf("Found tag of type %d with ID [%x %x %x %x %x %x %x]\n", 
			       tag_type, tid[0], tid[1], tid[2], tid[3], tid[4], tid[5], tid[6]);
		}
	}

	// Cycle through all types
	if(idle) {
		current_tag_type = (current_tag_type + 1) % 3;

		nfc_rfid_request_tag_id(nr, current_tag_type);
	}
}

int main(void) {
	// Create IP connection
	IPConnection ipcon;
	ipcon_create(&ipcon);

	// Create device object
	NFCRFID nr;
	nfc_rfid_create(&nr, UID, &ipcon);

	// Connect to brickd
	if(ipcon_connect(&ipcon, HOST, PORT) < 0) {
		fprintf(stderr, "Could not connect\n");
		return 1;
	}
	// Don't use device before ipcon is connected

	// Register state changed callback to function cb_state_changed
	nfc_rfid_register_callback(&nr,
	                           NFC_RFID_CALLBACK_STATE_CHANGED,
	                           (void (*)(void))cb_state_changed,
	                           &nr);

	// Start scan loop
	nfc_rfid_request_tag_id(&nr, NFC_RFID_TAG_TYPE_MIFARE_CLASSIC);

	printf("Press key to exit\n");
	getchar();
	nfc_rfid_destroy(&nr);
	ipcon_destroy(&ipcon); // Calls ipcon_disconnect internally
	return 0;
}

File: software/examples/java/ExampleScanForTags.java
import com.tinkerforge.IPConnection;
import com.tinkerforge.BrickletNFCRFID;

public class ExampleScanForTags {
	private static final String HOST = "localhost";
	private static final int PORT = 4223;
	private static final String UID = "XYZ"; // Change to your UID
	private static short currentTagType = 0;

	// Note: To make the example code cleaner we do not handle exceptions. Exceptions
	//       you might normally want to catch are described in the documentation
	public static void main(String args[]) throws Exception {
		IPConnection ipcon = new IPConnection(); // Create IP connection
		// Note: Declare nr as final, so the listener can access it
		final BrickletNFCRFID nr = new BrickletNFCRFID(UID, ipcon); // Create device object

		ipcon.connect(HOST, PORT); // Connect to brickd
		// Don't use device before ipcon is connected

		// Add state changed listener
		nr.addStateChangedListener(new BrickletNFCRFID.StateChangedListener() {
			public void stateChanged(short state, boolean idle) {
				try {
					if(idle) {
						currentTagType = (short)((currentTagType + 1) % 3);
						nr.requestTagID(currentTagType);
					}

					if(state == BrickletNFCRFID.STATE_REQUEST_TAG_ID_READY) {
						BrickletNFCRFID.TagID tagID = nr.getTagID();
						String s = "Found tag of type " + tagID.tagType + 
						           " with ID [" + Integer.toHexString(tagID.tid[0]);

						for(int i = 1; i < tagID.tidLength; i++) {
							s += " " + Integer.toHexString(tagID.tid[i]);
						}

						s += "]";
						System.out.println(s);
					}

				} catch(Exception e) {
					System.out.println(e);
				}
			}
		});

		// Start scan loop
		nr.requestTagID(BrickletNFCRFID.TAG_TYPE_MIFARE_CLASSIC);

		System.out.println("Press key to exit"); System.in.read();
		ipcon.disconnect();
	}
}
